crud/stats/init.lua

```diff
@@ -0,0 +1,291 @@
+--
+-- @function is_enabled
+--
+-- @treturn boolean Returns `true` or `false`.
+--
+function stats.is_enabled()
+    return internal.is_enabled == true
+end
+
+--- Initializes statistics registry, enables callbacks and wrappers.
+--
+--  If already enabled, do nothing.
+--
+-- @function enable
+--
+-- @treturn boolean Returns `true`.
+--
+function stats.enable()
+    if stats.is_enabled() then
+        return true
+    end
+
+    internal.is_enabled = true
+    registry.init()
+
+    return true
+end
+
+--- Resets statistics registry.
+--
+--  After reset collectors are the same as right
+--  after initial `stats.enable()`.
+--
+-- @function reset
+--
+-- @treturn boolean Returns true.
+--
+function stats.reset()
+    if not stats.is_enabled() then
+        return true
+    end
+
+    registry.destroy()
+    registry.init()
+
+    return true
+end
+
+--- Destroys statistics registry and disable callbacks.
+--
+--  If already disabled, do nothing.
+--
+-- @function disable
+--
+-- @treturn boolean Returns true.
+--
+function stats.disable()
+    if not stats.is_enabled() then
+        return true
+    end
+
+    registry.destroy()
+    internal.is_enabled = false
+
+    return true
+end
+
+--- Get statistics on CRUD operations.
+--
+-- @function get
+--
+-- @string[opt] space_name
+--  If specified, returns table with statistics
+--  of operations on space, separated by operation type and
+--  execution status. If there wasn't any requests of "op" type
+--  for space, there won't be corresponding collectors.
+--  If not specified, returns table with statistics
+--  about all observed spaces.
+--
+-- @treturn table Statistics on CRUD operations.
+--  If statistics disabled, returns `{}`.
+--
+function stats.get(space_name)
+    checks('?string')
+
+    if not stats.is_enabled() then
+        return {}
+    end
+
+    return registry.get(space_name)
+end
+
+-- Hack to set __gc for a table in Lua 5.1
+-- See https://stackoverflow.com/questions/27426704/lua-5-1-workaround-for-gc-metamethod-for-tables
+-- or https://habr.com/ru/post/346892/
+local function setmt__gc(t, mt)
+    local prox = newproxy(true)
+    getmetatable(prox).__gc = function() mt.__gc(t) end
+    t[prox] = true
+    return setmetatable(t, mt)
+end
+
+local function wrap_pairs_gen(build_latency, space_name, op, gen, param, state)
+    local total_latency = build_latency
+
+    -- If pairs() cycle will be interrupted with break,
+    -- we'll never get a proper obervation.
+    -- We create an object with the same lifespan as gen()
+    -- function so if someone break pairs cycle,
+    -- it still will be observed.
+    local observed = false
+
+    local gc_observer = setmt__gc({}, {
+        __gc = function()
+            if observed == false then
+                -- Do not call observe directly because it may
+                -- call fiber.yield and it is prohibited in gc.
+                fiber.new(registry.observe, total_latency, space_name, op, 'ok')
```

Comment:
Wow. Okay. We definitely need https://github.com/tarantool/tarantool/issues/5544.

Just curious: where the observe function yields?

Reply:
Well, I've studied metrics summary collector and it seems like it not yields in modern metrics versions (after https://github.com/tarantool/metrics/issues/235). But it can't be guaranteed that no new yields will be introduced (I've heard an idea to yield between quantiles flush if several quantiles used), and since yield in gc causes Tarantool to fail, I think it's better to guard it

Reply:
Thanks! It needs a bit more clarification (at least because it is introduced in the commit, where there is no metrics support). Something like 'a metrics collector may yield' or 'in some versions of metrics \<particular name of collector\> yields'.

Reply:
I think this one may be relevant (AFAIR, log yields): https://github.com/tarantool/metrics/blob/a23f8d49779205dd45bd211e21a1d34f26010382/metrics/collectors/shared.lua#L85

Reply:
Added a comment

Reply:
Thanks!